src/inmanta/data/__init__.py

```diff
@@ -402,41 +405,395 @@ def as_end_filter(
+            if not self.nullable:
+                raise ValueError("None is not a valid value")
+            else:
+                return None
+        if isinstance(value, self.base_type):
+            # It is as expected
+            return value
+        if issubclass(self.base_type, (str, int, bool)) and isinstance(value, (str, int, bool)):
+            # We can cast between those types
+            return self.base_type(value)
+        raise ValueError(f"{value} is not a valid value")
+
+    def get_accessor(self, column_name: str, table_prefix: Optional[str] = None) -> str:
+        """
+        return the sql statement to get this column, as used in filter and other statements
+        """
+        table_prefix_value = "" if table_prefix is None else table_prefix + "."
+        return table_prefix_value + column_name
+
+    def coalesce_to_min(self, value_reference: str) -> str:
+        """If the order by column is nullable, coalesce the parameter value to the minimum value of the specific type
+        This is required for the comparisons used for paging, because comparing a value to
+        NULL always yields NULL.
+        """
+        if self.nullable:
+            if self.base_type == datetime.datetime:
+                return f"COALESCE({value_reference}, to_timestamp(0))"
+            elif self.base_type == bool:
+                return f"COALESCE({value_reference}, FALSE)"
+            elif self.base_type == int:
+                return f"COALESCE({value_reference}, 0)"
+            else:
+                return f"COALESCE({value_reference}, '')"
+        return value_reference
+
+
+class ForcedStringCollumn(ColumnType):
+    """A string that is explicitly cast to a specific string type"""
+
+    def __init__(self, forced_type: str) -> None:
+        super().__init__(base_type=str, nullable=False)
+        self.forced_type = forced_type
+
+    def get_accessor(self, column_name: str, table_prefix: Optional[str] = None) -> str:
+        """
+        return the sql statement to get this column, as used in filter and other statements
+        """
+        return super().get_accessor(column_name, table_prefix) + "::" + self.forced_type
+
+
+StringColumn = ColumnType(base_type=str, nullable=False)
+DateTimeColumn = ColumnType(base_type=datetime.datetime, nullable=False)
+IntColumn = ColumnType(base_type=int, nullable=False)
+TextColumn = ForcedStringCollumn("text")
+UUIDColumn = ColumnType(base_type=uuid.UUID, nullable=False)
+
+
+class DatabaseOrderV2(ABC):
+    """
+    Helper API for handling database order and filtering
+
+    This class defines the consumer interface,
+
+    It is made into a separate type, to make it very explicit what is exposed externally, to limit feature creep
+    """
+
+    @abstractmethod
+    def as_filter(
+        self,
+        offset: int,
+        column_value: Optional[PRIMITIVE_SQL_TYPES] = None,
+        id_value: Optional[PRIMITIVE_SQL_TYPES] = None,
+        start: Optional[bool] = True,
```

Comment:
```suggestion
        start: bool = True,
```
I don't think this is expected to be `None`, is it? Same for children.

Reply:
done